### ISSUE TITLE
[JOSS-Review] Much better convincing Readme

### ISSUE THREAD

**Issue by simonprovost** [rvandewater/recipys#26]
Hi folks!

The readme is pretty great, but I believe it might be more useful with certain visualisations.  If it is not desired, that is perfectly great; I am offering some help; take it or leave it 🫡

1) Begin by aligning the repository's name at the top and incorporating a subtitle to create a much better look with the logo centred (eyes do not need to go left to right now, start middle and keep left). Additionally, this also assume centering the badges.

 2) Incorporate widely used `pip` (already-there of course) and `uv` installation methods, and for the editable version, wrap it in a collapsable head-content to enhance the professional Readme appearance.

3) I suggest incorporating a visualisation that outlines Recipes in a step-by-step diagram format. This would allow for both the `core concepts` and `backend` subsections to be described below the visual rather than in two sections.  I quickly created a diagram in Figma as an example. If you find it interesting, I can share the Figma file for you to fork and modify with more accurate information. I'm happy to help with that. Please see the preview below.

4) Relocate your example section to the Quick start one and rename it to Getting started to optimise space and reduce the total number of sections.

5) Combine your contributions and License into one.

6) Eliminate the emoji from your `Paper` section to maintain consistency with the others, and consider renaming it to a more standard title such as `How To Cite Us?` or something similar like `How to Cite ReCiPes?` at the very least.

Overall I do not mind how the readme become, this is your OSS, but it deserve a bit more attention now that the repo is stable and the paper too! If against all of the above, feel free to comment how, please.

Cheers 🥳

———

# Could look like that with some extra-taken-initiatives so do not just copy & paste ⬇️

<div align="center">
  <img src="https://github.com/rvandewater/ReciPies/blob/development/docs/figures/recipies_logo.svg?raw=true" 
       alt="ReciPies logo" height="300">
</div>

<h1 align="center">ReciPies</h1>
<p align="center"><em>A Novel Declarative, reusable, reproducible preprocessing pipelines</em></p>

<p align="center">
  <a href="https://github.com/rvandewater/ReciPies/actions/workflows/ci.yml"><img src="https://github.com/rvandewater/ReciPies/actions/workflows/ci.yml/badge.svg" alt="CI"></a>
  <img src="https://img.shields.io/badge/platform-linux--64%20|%20win--64%20|%20osx--64-lightgrey" alt="Platform">
  <a href="LICENSE"><img src="https://img.shields.io/badge/license-MIT-green" alt="License: MIT"></a>
  <a href="https://pypi.python.org/pypi/recipies/"><img src="https://img.shields.io/pypi/v/recipies.svg" alt="PyPI version"></a>
  <a href="https://pypi.python.org/pypi/recipies/"><img src="https://img.shields.io/pypi/pyversions/recipies.svg" alt="Python Versions"></a>
  <a href="https://pepy.tech/project/recipies"><img src="https://pepy.tech/badge/recipies" alt="Downloads"></a>
  <a href="http://arxiv.org/abs/2306.05109"><img src="https://img.shields.io/badge/arXiv-2306.05109-b31b1b.svg" alt="arXiv"></a>
</p>

Modern machine learning (ML) workflows live or die by their data‑preprocessing steps, yet in Python—a language with a 
rich ecosystem for data science and ML—these steps are often scattered across ad‑hoc scripts or opaque Scikit-Learn 
(sklearn) snippets that are hard to read, audit, or reuse. `ReciPies` provides a concise, human‑readable, and fully 
reproducible way to declare, execute, and share preprocessing pipelines, adhering to Configuration as Code principles. 
It lets users describe transformations as a recipe made of ordered *steps* (e.g., imputing, encoding, normalizing) 
applied to variables identified by semantic roles (predictor, outcome, ID, time stamp, etc.). Recipes can be *prepped* 
(trained) once, *baked* many times, and cleanly separated between training and new data—preventing data leakage by 
construction. Under the hood, `ReciPies` targets both Pandas and Polars backends for performance and flexibility, and 
it is easily extensible: users can register custom steps with minimal boilerplate. Each recipe is serializable to 
JSON/YAML for provenance tracking, collaboration, and publication, and integrates smoothly with downstream modeling 
libraries. Packaging preprocessing as clear, declarative objects, `ReciPies` lowers the cognitive load of feature 
engineering, improves reproducibility, and makes methodological choices explicit, benefiting individual researchers, 
engineering teams, and peer reviewers alike.

The backend can either be [Polars](https://github.com/pola-rs/polars) or [Pandas](https://github.com/pandas-dev/pandas) dataframes. 

The operation of this package is inspired by the R-package [recipes](https://recipes.tidymodels.org/). 

Please check the [documentation](rvandewater.github.io/ReciPies/) for more details.

### Using `pip`
```bash
pip install recipies
```

### Using `uv`
```bash
uv add recipies
```

<details>
<summary><strong>Developer / Editable install</strong></summary>

#### From source (editable)
```bash
# with conda (optional)
conda env update -f environment.yml
conda activate ReciPies

# or with uv venv
# uv venv && source .venv/bin/activate

pip install -e .
# alternatively with uv:
# uv pip install -e .
```

</details>

## Getting Started

Here's a simple example of using ReciPies:

```python
import polars as pl, numpy as np
from datetime import datetime, MINYEAR
from recipies import Ingredients, Recipe
from recipies.selector import all_numeric_predictors, all_predictors
from recipies.step import StepSklearn, StepHistorical, Accumulator, StepImputeFill
from sklearn.impute import MissingIndicator

rand_state = np.random.RandomState(42)

timecolumn = pl.concat([
  pl.datetime_range(datetime(MINYEAR, 1, 1, 0), datetime(MINYEAR, 1, 1, 5), "1h", eager=True),
  pl.datetime_range(datetime(MINYEAR, 1, 1, 0), datetime(MINYEAR, 1, 1, 3), "1h", eager=True)
])

df = pl.DataFrame({
  "id": [1] * 6 + [2] * 4,
  "time": timecolumn,
  "y": rand_state.normal(size=(10,)),
  "x1": rand_state.normal(loc=10, scale=5, size=(10,)),
  "x2": rand_state.binomial(n=1, p=0.3, size=(10,)),
  "x3": pl.Series(["a", "b", "c", "a", "c", "b", "c", "a", "b", "c"], dtype=pl.Categorical),
  "x4": pl.Series(["x", "y", "y", "x", "y", "y", "x", "x", "y", "x"], dtype=pl.Categorical),
})

df = df.with_columns(
  pl.when(pl.int_range(pl.len()).is_in([1, 2, 4, 7]))
  .then(None)
  .otherwise(pl.col("x1"))
  .alias("x1")
)

df2 = df.clone()

ing = Ingredients(df)
rec = Recipe(
  ing,
  outcomes=["y"],
  predictors=["x1", "x2", "x3", "x4"],
  groups=["id"],
  sequences=["time"]
)

rec.add_step(StepSklearn(MissingIndicator(features="all"), sel=all_predictors()))
rec.add_step(StepImputeFill(sel=all_predictors(), strategy="forward"))
rec.add_step(StepHistorical(sel=all_predictors(), fun=Accumulator.MEAN, suffix="mean_hist"))

# train the pipeline on the training data
df = rec.prep()

# apply to new data (e.g., test set)
df2 = rec.bake(df2)
```

For more (notebooks) examples, see the `examples/` directory; performance comparisons are in `benchmarks/`.

## `ReciPies` In A Nusthell

<img width="4708" height="1752" alt="Image" src="https://github.com/user-attachments/assets/d143ca6d-76b9-4396-ab85-e46e071415fa" />

### `Core Concepts` & `Backends Supported`

`Ingredients`: Wrapper around DataFrames that maintains column role information, ensuring data semantics are preserved during transformations. —— `Recipe`: A collection of ordered processing steps applied to an `Ingredients` object to create reproducible data pipelines. —— `Step`: Individual data transformation operations that understand column roles and work with both Polars and Pandas backends. —— `Selector`: Utilities for selecting columns based on their roles or other criteria (e.g., “all predictors”).

`ReciPies` supports both **Polars** and **Pandas** backends:

- **Polars** — high-performance DataFrame library with lazy evaluation.
- **Pandas** — traditional DataFrame library with extensive ecosystem support.

The package can automatically detect the backend and provides a consistent API regardless of the underlying DataFrame implementation.

## Contributing & License

Contributions are welcome! Please open an issue or submit a pull request on the [GitHub repository](https://github.com/rvandewater/ReciPies).  
By contributing, you agree that your contributions will be licensed under the **MIT License**.

- **License:** See [LICENSE](https://github.com/rvandewater/ReciPies/blob/main/LICENSE) for details.
- **Code of Conduct / Guidelines:** See [CODE_OF_CONDUCT.md](https://github.com/rvandewater/ReciPies?tab=coc-ov-file) and [CONTRIBUTING.md](https://github.com/rvandewater/ReciPies?tab=contributing-ov-file).

## How to Cite `ReciPies`

If you use this code in your research, please cite the following publication (ReciPies is used extensively to build a customisable preprocessing pipeline; a standalone paper is in preparation):

```
@inproceedings{vandewaterYetAnotherICUBenchmark2024,
  title = {Yet Another ICU Benchmark: A Flexible Multi-Center Framework for Clinical ML},
  shorttitle = {Yet Another ICU Benchmark},
  booktitle = {The Twelfth International Conference on Learning Representations},
  author = {van de Water, Robin and Schmidt, Hendrik Nils Aurel and Elbers, Paul and Thoral, Patrick and Arnrich, Bert and Rockenschaub, Patrick},
  year = {2024},
  month = oct,
  urldate = {2024-02-19},
  langid = {english},
}
```

_This paper can also be found on arxiv: https://arxiv.org/pdf/2306.05109.pdf_

**Comment by rvandewater** [comment]
@simonprovost I have added a diagram to show how recipies works @simonprovost. I do see your other suggestions and I have incorporated many but for some I don't really see how this with improve the clarity of the readme which is supposed to be rather a quick piece of information.

**Comment by simonprovost** [comment]
> [@simonprovost](https://github.com/simonprovost) I have added a diagram to show how recipies works [@simonprovost](https://github.com/simonprovost). I do see your other suggestions and I have incorporated many but for some I don't really see how this with improve the clarity of the readme which is supposed to be rather a quick piece of information.

Definitely agreed!  That's why I prefer this to be an open conversation about how to improve the clarity and effectiveness of your readme, which is completely great.  Could you please make a list of what has "changed" overall —— and a branch to which I could viz. the updated readme or if it's merge on your dev/main track just let me know of that.

By the way, if the readme is getting too long, perhaps the most practical option would be to include a tab in your documentation that explains Recipes in further detail, e.g. with a diagram, the backends/core concepts explanations etc. Have a thought, not requesting it but just sharing an extra option.

Note that I am more here to ensure that the best line of action has been taken, and the actual outcome of your OSS is your call of course

Let me know:)

**Comment by rvandewater** [comment]
Hi @simonprovost  
- I've added a diagram showing how ReciPies works
- Added uv documentation
- Incorporated most of your smaller suggestions
Let me know if you have any issues with the current state of the readme :).

**Comment by simonprovost** [comment]
Readme is brillant! At the very least, stable enough to grow over time, thanks for your promptness @rvandewater, very much appreciated!
